sdk/tests/core/test_az_span.c

```diff
@@ -129,24 +130,377 @@ static void test_az_span_is_content_equal(void** state)
+  az_span_atox_return_errors_helper(AZ_SPAN_FROM_STR("--123"));
+  az_span_atox_return_errors_helper(AZ_SPAN_FROM_STR("-+123"));
+  az_span_atox_return_errors_helper(AZ_SPAN_FROM_STR("  -1-"));
+  az_span_atox_return_errors_helper(AZ_SPAN_FROM_STR("- INFINITY"));
+  az_span_atox_return_errors_helper(AZ_SPAN_FROM_STR("- 0"));
+  az_span_atox_return_errors_helper(AZ_SPAN_FROM_STR("+ 1"));
+  az_span_atox_return_errors_helper(AZ_SPAN_FROM_STR("1.-e3"));
+  az_span_atox_return_errors_helper(AZ_SPAN_FROM_STR("1.-e/3"));
+  az_span_atox_return_errors_helper_exclude_double(AZ_SPAN_FROM_STR("1.23"));
+  az_span_atox_return_errors_helper_exclude_double(AZ_SPAN_FROM_STR("1.23e3"));
+  az_span_atox_return_errors_helper_exclude_double(AZ_SPAN_FROM_STR("99999999999999999999"));
+  az_span_atox_return_errors_helper_exclude_double(AZ_SPAN_FROM_STR("999999999999999999999"));
+  az_span_atox_return_errors_helper_exclude_double(AZ_SPAN_FROM_STR("18446744073709551616"));
+  az_span_atox_return_errors_helper_exclude_double(AZ_SPAN_FROM_STR("-18446744073709551616"));
 }
 
 static void az_span_atou32_test(void** state)
 {
   (void)state;
-  az_span number = AZ_SPAN_FROM_STR("1024");
   uint32_t value = 0;
 
-  assert_return_code(az_span_atou32(number, &value), AZ_OK);
+  assert_return_code(az_span_atou32(AZ_SPAN_FROM_STR("0"), &value), AZ_OK);
+  assert_int_equal(value, 0);
+  assert_return_code(az_span_atou32(AZ_SPAN_FROM_STR("1024"), &value), AZ_OK);
   assert_int_equal(value, 1024);
+  assert_return_code(az_span_atou32(AZ_SPAN_FROM_STR("2147483647"), &value), AZ_OK);
+  assert_int_equal(value, 2147483647);
+  assert_return_code(az_span_atou32(AZ_SPAN_FROM_STR("4294967295"), &value), AZ_OK);
+  assert_int_equal(value, 4294967295);
+
+  assert_return_code(az_span_atou32(AZ_SPAN_FROM_STR("-123"), &value), AZ_OK);
+  assert_int_equal(value, 4294967295 - 123 + 1);
+  assert_return_code(az_span_atou32(AZ_SPAN_FROM_STR("-2147483648"), &value), AZ_OK);
+  assert_int_equal(value, 4294967295 - 2147483648 + 1);
+  assert_return_code(az_span_atou32(AZ_SPAN_FROM_STR("-4294967295"), &value), AZ_OK);
+  assert_int_equal(value, 4294967295 - 4294967295 + 1);
+  assert_return_code(az_span_atou32(AZ_SPAN_FROM_STR("-4294967296"), &value), AZ_OK);
+  assert_int_equal(value, 4294967295 - 4294967296 + 1);
+  assert_return_code(az_span_atou32(AZ_SPAN_FROM_STR("4294967296"), &value), AZ_OK);
+  assert_int_equal(value, 4294967295 - 4294967296 + 1);
+  assert_return_code(az_span_atou32(AZ_SPAN_FROM_STR("42949672950"), &value), AZ_OK);
+  assert_int_equal(value, 4294967295 - 10 + 1);
+  assert_return_code(az_span_atou32(AZ_SPAN_FROM_STR("9223372036854775807"), &value), AZ_OK);
+  assert_int_equal(value, 4294967295);
+  assert_return_code(az_span_atou32(AZ_SPAN_FROM_STR("9223372036854775808"), &value), AZ_OK);
+  assert_int_equal(value, 0);
+  assert_return_code(az_span_atou32(AZ_SPAN_FROM_STR("18446744073709551615"), &value), AZ_OK);
+  assert_int_equal(value, 4294967295);
```

Comment:
`sscanf` has somewhat odd behavior when it comes to parsing string to integer types, like atou32, when it comes to out-of-bounds numbers (negative numbers or numbers greater than uint32.Max -> 4294967295). 

Are we OK with this behavior, @JeffreyRichter, or should we only use `sscanf` **only** for `atod` and keep our own implementation for `ato[u|i][32|64]`?

Reply:
This issue got addressed by using a more compatible/compiler agnostic format string (`%llu` instead of `%lu`).

From Charlie:
> the "%lu" specifier is for an unsigned long, which is 64 bit on gcc/clang(in gcc mode) and 32bit on msvc